Create/Edit page for Webhook secrets

jhadvig opened the pull request:
Added button with `Create New` dropdown button, which (for now, rest will be added in a follow up) contains:
- Create Webhook Secret
- Create via YAML

![1](https://user-images.githubusercontent.com/1668218/41368803-440c6364-6f43-11e8-87c6-0518a66f1b18.png)

When creating the secret Im adding all the values to the `stringData` which will take care of encoding the data.

![2](https://user-images.githubusercontent.com/1668218/41368927-a8867d20-6f43-11e8-91bd-1db2ae8d3546.png)

I've also added following actions for individual secrets:
- Modify Labels
- Modify Annotations
- Duplicate Secret
- Edit Secret
- Delete Secret

![3](https://user-images.githubusercontent.com/1668218/41368958-c361877a-6f43-11e8-9ff0-741a945fa477.png)

PTAL

spadgett commented on frontend/package.json:
```diff
@@ -68,6 +68,7 @@
     "plotly.js": "1.28.x",
     "prop-types": "15.6.x",
     "react": "16.x",
+    "react-ace": "6.1.x",
```
We're already using Ace editor elsewhere (edit YAML). I think we should be able to use Ace here without the new dependency. Can you check?

https://github.com/openshift/console/blob/master/frontend/public/components/edit-yaml.jsx

spadgett replied:
Sorry I see you comment on this in the description. I think we should discuss further. How large is the library?

jhadvig replied:
from what I see in the node_modules it's 2MB. Just for comparision the `brace` has 8.5MB

ggreer replied:
Don't look at node_modules size. Use `yarn run analyze` to see the size of the built js.

kans replied:
react-ace is a thin wrapper around brace:

https://github.com/securingsincity/react-ace/blob/master/package.json#L63:
```javascript
  "dependencies": {
    "brace": "^0.11.0",
   ...
},
```

https://github.com/securingsincity/react-ace/blob/master/src/ace.js#L1:
```javascript
import ace from 'brace'
```

`react-ace` itself is 46.47 KB.  Previously we were async loading `brace` only when needed. Since secret.jsx now imports `brace` (747KB), our main bundle of vendored deps has increased in size by ~40%.  

Are there any interactions here that aren't 90% copy/paste - ie, why do we need a text editor at all?  If you do need one, you should use `AsyncComponent` to load it only when needed since the vast majority of users will not edit a secret.

spadgett commented:
cc @tlwu2013

robszumski commented:
Throwing out this multi-create button pattern that we have for OLM:

![image](https://user-images.githubusercontent.com/755540/41056115-f9e08f7a-6977-11e8-8da1-4535dcb78425.png)

What does @openshift/team-ux-review think of offloading the selection type to this button, which will reduce one of the fields on the screen, plus give you an option to jump to a raw YAML editor. Using a dropdown to switch this will alter the page a ton, plus you have no way to get back.

```
Create Source Secret...
Create Key/Value Secret...
Create Image Secret...
Create Webhook Secret...
Create via YAML...
```

Also, I think we need some microcopy work here and none of those phrases mean anything to me:
 - "source" and "image" are very generic and don't match a Kubernetes-ism that I may be looking for
 - I renamed "generic" -> "key/value" above
 - Should "Image" be "ImagePullSecret" (direct) or "Pull Secret" (friendlier)

kans commented:
Do you mean `Opaque` instead of `Generic`?  If so, lets use the k8s term.

spadgett commented:
@jhadvig I'd suggest doing these one at a time in separate PRs, maybe starting with opaque secrets. I think we'll be able to get these reviewed and merged quicker that way.

I don't know if the name/value editor will work as-is for generic secrets since they often have multiline values (e.g. certificates and keys).

It would be good to structure this in such a way that we can use the same components for editing secrets, too.

cshinn commented:
I think @robszumski's idea makes sense. Especially if that selection will drastically change the form underneath. As for terminology, sticking with the K8s standards is probably a good idea

jhadvig commented:
@spadgett I've update the PR based on the comments from @robszumski 
I've replaced the `Create Secret` button with `Create New` dropdown button, which (for now, rest will be added in a follow up) contains:
- Create Webhook Secret
- Create via YAML

![1](https://user-images.githubusercontent.com/1668218/41368803-440c6364-6f43-11e8-87c6-0518a66f1b18.png)

When creating the secret Im adding all the values to the `stringData` which will take care of encoding the data.

![2](https://user-images.githubusercontent.com/1668218/41368927-a8867d20-6f43-11e8-91bd-1db2ae8d3546.png)

I've also added following actions for individual secrets:
- Modify Labels
- Modify Annotations
- Duplicate Secret
- Edit Secret
- Delete Secret

![3](https://user-images.githubusercontent.com/1668218/41368958-c361877a-6f43-11e8-9ff0-741a945fa477.png)

PTAL

cshinn commented:
@jhadvig the dropdown panel on the "Create New" button looks a bit misaligned. Looks great other than that

beanh66 commented:
@jhadvig Thanks this looks good, I just have a few comments: 

- Can we fix the alignment as Chris mentioned?
- The dropdown can just say "Create" with the caret 
- The options can then be "Secret from YAML" and "Webhook Secret" 
- Options should be alphabetical 
- The padding on the right should match the padding on the left if possible

jhadvig commented on frontend/public/components/secrets/create-secret.jsx:
```diff
@@ -0,0 +1,170 @@
+        case 'source':
+        case 'image':
+        case 'generic':
+        case 'webhook':
```
No.. those are just subform options that console will eventually exist

alecmerdler replied:
Right now the `case` statements above will fall through to the `'webhook'` option. I would suggest YAGNI here and remove the `switch` block altogether.

jhadvig replied:
True, but the switch will be there once we add all the subforms options. Just wanted to have all the options on mind. But don't have any strong preferences regarding changing it to `if` statement :)

jhadvig commented on frontend/public/components/factory/list-page.jsx:
```diff
@@ -246,7 +246,19 @@ export const ListPage = props => {
     } catch (unused) { /**/ }
   }
-  const createProps = createHandler ? {onClick: createHandler} : {to: href};
+  let createProps;
```
Not sure how the [FireMan_](https://github.com/openshift/console/blob/master/frontend/public/components/factory/list-page.jsx#L163) will evaluate the truly the Dropdown component, since the [ListPage component](https://github.com/openshift/console/blob/master/frontend/public/components/factory/list-page.jsx#L249) sets `createProps` either with **onClick** or **to** keys, there is no **items** key set, so there is no way [Dropdown](https://github.com/openshift/console/blob/master/frontend/public/components/factory/list-page.jsx#L165) will be rendered. Or maybe I'm missing something

alecmerdler replied:
See the ["Create New" button for the `ClusterServiceVersion-v1` detail view](https://teamui.console.team.coreos.systems/k8s/ns/alm-dev/clusterserviceversion-v1s/etcdoperator.v0.9.2) and [its implementation](https://github.com/openshift/console/blob/master/frontend/public/components/cloud-services/clusterserviceversion-resource.tsx#L97).

jhadvig replied:
But the [implementation](https://github.com/openshift/console/blob/master/frontend/public/components/cloud-services/clusterserviceversion-resource.tsx#L97) counts with setting the `items` which is missing in the [ListPage component](https://github.com/openshift/console/blob/master/frontend/public/components/factory/list-page.jsx#L249).

alecmerdler replied:
I see what you mean. Try this:
1. Allow passing `createProps` as a prop to `ListPage`
2. Change [this line](https://github.com/openshift/console/blob/master/frontend/public/components/factory/list-page.jsx#L249) to: 
```javascript
const createProps = createProps || (createHandler ? {onClick: createHandler} : {to: href});
```
3. Change `SecretsPage` to this:
```javascript
const SecretsPage = props => {
  const createItems = {
    // source: 'Create Source Secret',
    // image: 'Create Image Pull Secret',
    // generic: 'Create Key/Value Secret',
    webhook: 'Webhook Secret',
    yaml: 'Secret from YAML',
  };

  const createProps = {
    items: createItems,
    createLink: (type) => `/k8s/ns/${props.obj.metadata.namespace}/secrets/${type === 'yaml' ? 'new' : type}`
  };

  return <ListPage ListComponent={SecretsList} canCreate={true} rowFilters={filters} createButtonText="Create" createProps={createProps} {...props} />;
};
```
I tried this locally and it works.

jhadvig commented:
@beanh66 
> Options should be alphabetical
I understand the point but thought that options that will render an actual form should be first and the YAML editor should be the last one and by that somehow logically group them.
Thoughts ?

beanh66 commented:
@jhadvig Got it, that makes sense to me on the order.

spadgett commented:
It looks like we need to update the integration tests for this change.

jhadvig commented on frontend/public/components/secret.jsx:
```diff
@@ -16,7 +16,19 @@ data:
+  Cog.factory.ModifyLabels,
+  Cog.factory.ModifyAnnotations,
+  (kind, obj) => ({
+    label: `Duplicate ${kind.label}...`,
```
I've seen that some of the sereources have the option to duplicate, but to be honest I wasnt sure if it should be available for secrets. Will put it away

jhadvig commented on frontend/public/components/secrets/create-secret.jsx:
```diff
@@ -0,0 +1,163 @@
+              <label className="control-label">Secret Name</label>
+              <div className="modal-body__field">
+                <input className="form-control" type="text" onChange={this.changeSecretName} value={this.state.secret.metadata.name} required id="test--subject-name" />
+                <p className="help-block text-muted">Unique name of the new secret.</p>
```
Found we are using the `text-muted` in other place, agreed that `help-block` should be enough

jhadvig commented on frontend/public/components/secrets/create-secret.jsx:
```diff
@@ -0,0 +1,163 @@
+import * as _ from 'lodash-es';
```
you mean to replace `.jsx` with `.tsx` ?

jhadvig commented on frontend/public/components/secrets/create-secret.jsx:
```diff
@@ -0,0 +1,163 @@
+    constructor (props) {
+      super(props);
+      const kind = 'Secret';
+      const existingData = _.pick(props.obj, ['metadata.name', 'metadata.namespace', 'data']);
```
agreed, I've picked those two mainly because of the `duplicate` action but since we've nuked that options I will put the whole `metadata` field there.

jhadvig commented:
@spadgett I've updated the PR:
- rewrote the `subforms` and `create-secret` into typescript - had to update `button-bar.jsx` because of that 
- updated the intergration-tests
- only when creating webhook secret the form will be displayed, otherwise the YAML editor. The same applies for editing secrets, there only `WebHookSecretKey` key has to be present in the `data`.
 
Although I'm not sure why the tests are failing :-/

spadgett commented:
Ignore the ci/prow/* statuses.

Linting is failing, though:

```
go/src/github.com/openshift/console/frontend/public/components/utils/dropdown.jsx
  error Unexpected string concatenation prefer-template
```

spadgett commented:
You can always use `yarn run lint` to run the linter locally

jhadvig commented:
/retest

jhadvig commented:
@spadgett sorry for that ... thought that interpolating won't `id="${itemKey}-link"` wont cause lint to scream, but apparently `id={`${itemKey}-link`} ` is the right form

spadgett commented on frontend/integration-tests/tests/crud.scenario.ts:
```diff
@@ -85,7 +85,12 @@ describe('Kubernetes resource CRUD operations', () => {
 
       it('displays a YAML editor for creating a new resource instance', async() => {
-        await crudView.createYAMLButton.click();
+        if (kind === 'Secret') {
```
Instead of checking `kind`, can we check for the existence of `#item-create`? Then if we add similar buttons to other pages, the tests don't need to change.

spadgett commented on frontend/public/components/secret.jsx:
```diff
@@ -16,7 +16,22 @@ data:
 
-const menuActions = Cog.factory.common;
+const editInYaml = obj => {
+  if (obj.type === 'Opaque' && _.has(obj.data, 'WebHookSecretKey') && Object.keys(obj.data).length === 1) {
```
I would use

```js
if (_.has(obj, 'data.WebHookSecretKey') && _.size(obj.data) === 1) {
```

The whole function could be simplified to

```js
// Edit in YAML if not editing a webhook secret with one key.
const editInYaml = obj => !_.has(obj, 'data.WebHookSecretKey') || _size(obj.data) !== 1;
```

We should make the string `WebHookSecretKey` a const defined once somewhere that we reuse to avoid typos.

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+      secret: secret,
+      inProgress: false,
+      type: secret.type,
+      stringData: _.mapValues(secret.data, (v) => window.atob(v)),
```
I think this can be

```js
stringData: _.mapValues(secret.data, window.atob),
```

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+      type: secret.type,
+      stringData: _.mapValues(secret.data, (v) => window.atob(v)),
+    };
+    this.onNameChanged = this.onNameChanged.bind(this);
```
You should bind `onDataChanged` here, too

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    e.preventDefault();
+    const { kind, metadata } = this.state.secret;
+    this.setState({ inProgress: true });
+    const newSecret = _.assign({}, this.state.secret, {stringData: this.state.stringData});
```
Does this include both `data` and `stringData`? We should only include `stringData`

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    ).then(
+      () => {
+        this.setState({inProgress: false});
+        if (metadata.namespace) {
```
We should always have a namespace at this point for both create and update, right?

jhadvig replied:
yes, we should

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    const title = `${this.props.titleVerb} ${_.upperFirst(this.state.secretType)} Secret`;
+    const { saveButtonText } = this.props;
+
+    let explanation = 'Webhook secret allow you to authenticate a webhook trigger.';
```
'Webhook secrets allow you to authenticate a webhook trigger.'

benjaminapetersen replied:
`const explanation;`
`const subform`;

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    const { saveButtonText } = this.props;
+
+    let explanation = 'Webhook secret allow you to authenticate a webhook trigger.';
+    let subform = <WebHookSecretSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData} />;
```
It would be nice if the base edit component didn't need to know about the specific secret types. I wonder if we can't use a higher-order component here:

https://reactjs.org/docs/higher-order-components.html#use-hocs-for-cross-cutting-concerns

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+
+const BaseEditSecret = connect(null, {setActiveNamespace: UIActions.setActiveNamespace})(
+  (props: BaseEditSecretProps_) => <BaseEditSecret_ {...props} />
+);
```
Can't this just be...

```js
const BaseEditSecret = connect(null, {setActiveNamespace: UIActions.setActiveNamespace})(BaseEditSecret_);
```

jhadvig replied:
No the `props` need to be passed explicitly from parent to child otherwise, yarn will scream that some properties does not exist:
> Property 'fixed' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<any, never>, ComponentState, never>...'.

spadgett replied:
Then I think we must have declared the types incorrectly somehow. We should figure out what's wrong.

jhadvig replied:
The issue is that the redux part of the code is in jsx an therefor is not typed, eg. [ui-actions.js](https://github.com/openshift/console/blob/master/frontend/public/ui/ui-actions.js). I think until we type the redux part of the then codebase, which means rewriting it to TS, until then we need this as a workaround.

spadgett commented on frontend/public/components/secrets/subforms.tsx:
```diff
@@ -0,0 +1,49 @@
+const generateSecret = () => {
+  // http://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
+  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
+  return s4()+s4()+s4()+s4();
```
nit: spaces before and after `+`

spadgett commented on frontend/public/components/utils/button-bar.jsx:
```diff
@@ -21,12 +21,12 @@ const InfoMessage = ({message}) => <div className="alert alert-info"><span class
+    {props.errorMessage && <ErrorMessage message={props.errorMessage} />}
+    {injectDisabled(props.children, props.inProgress)}
+    {props.inProgress && <LoadingInline />}
+    {props.infoMessage && <InfoMessage message={props.infoMessage} />}
```
It's not clear to me why this change was needed?

jhadvig replied:
There was an issue with the required properties for the ButtonBar
>TS2322: Type '{ children: Element[]; errorMessage: any; inProgress: boolean; }' is not assignable to type 'IntrinsicAttributes & { children: any; className: any; errorMessage: any; infoMessage: any; inPro...'.
  Type '{ children: Element[]; errorMessage: any; inProgress: boolean; }' is not assignable to type '{ children: any; className: any; errorMessage: any; infoMessage: any; inProgress: any; }'.
    Property 'className' is missing in type '{ children: Element[]; errorMessage: any; inProgress: boolean; }'.
Version: typescript 2.7.2
Time: 12738ms

It was cause by the listed params in the class definition
> export const ButtonBar = ({children, className, errorMessage, infoMessage, inProgress}) => {
...
}

Because of listing the parameters and not just setting the to `props`, all the listed params will be required, even if `propTypes` will declare that they are not.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+
+const determineSecretTypeAbstraction = (secret) => {
+  const { data } = secret;
+  if (_.has(data, 'WebHookSecretKey')) {
```
Style nit, but I think you will see something like:

```JavaScript 
return _.has(data, 'WebHookSecretKey') 
  ? 'webhook'
  : 'generic';
```
Throughout the app.  If possible I think we should try to conform.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretState_> {
+  constructor (props) {
+    super(props);
+    let existingData = _.pick(props.obj, ['metadata', 'data']);
```
I think you can `const existingData`.  It isn't changing.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    this.save = this.save.bind(this);
+  }
+  onDataChanged (secretsData) {
+    let stringData = {...this.state.stringData};
```
`const`

benjaminapetersen replied:
Also, you are defining `stringData` then immediately setting it as something else. 

Could this just be:


```JavaScript 
this.setState({... secretsData});
```

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    this.setState({stringData});
+  }
+  onNameChanged (event) {
+    let secret = {...this.state.secret};
```
es6:

```JavaScript
this.setState({
  ...this.state.secret,
  metadata: {
    name: event.target.value
  }
});
```

jhadvig replied:
That will just set the `name` in the `metadata` field, everything else will be removed when set. Tried:
> this.setState({
  ...this.state.secret,
  metadata: {
    name: event.target.value
    ...this.state.secret.namespace,
  }
});

but thats not setting the `name` at all

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+    (this.props.isCreate
+      ? k8sCreate(ko, newSecret)
+      : k8sUpdate(ko, newSecret, metadata.namespace, newSecret.metadata.name)
+    ).then(
```
Mind collapsing some parens?  Ex:  `.then(() => {`.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+</Firehose>;
+
+/* eslint-disable no-undef */
+export type BaseEditSecretState_ = {
```
Is the trailing underscore common?  I haven't seen previously.

jhadvig replied:
It is, at least we use it in our [codebase](https://github.com/openshift/console/blob/master/frontend/public/components/RBAC/bindings.jsx#L343)

benjaminapetersen replied:
Huh, ok neat.

spadgett replied:
Not typically for the types, though?

jhadvig replied:
Probably not. If anything we could rename:
- `BaseEditSecret`    ->  `BaseEditSecretWrapper`
- `BaseEditSecret_`  ->  `BaseEditSecret`

spadgett replied:
I think the names are good. Just wondering if we should have the underscore on the type definitions

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+import { SafetyFirst } from '../safety-first';
+import { WebHookSecretSubform } from './subforms';
+
+const determineSecretTypeAbstraction = (secret) => {
```
Might be fine to name this `determineSecretType()` or just `secretType()`.

jhadvig replied:
Secrets are a kind already have `type` field, and all the values that could return from that function will be just an type abstractions:
- webhook
- source
- image
- key/value

benjaminapetersen replied:
Ok makes sense.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+  kind?: string,
+  isCreate: boolean,
+  titleVerb: string,
+  setActiveNamespace: any,
```
`setActiveNamespace` must be a function, correct?  I believe this can be typed.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,155 @@
+  error?: any,
+};
+
+export type BaseEditSecretProps_ = {
```
Can we eliminate the `any` types?
For example:
```
 obj: K8sResourceKind 
// ...
metadata: {
  name: string,
  // etc.
}
```
I think [this file](https://github.com/alecmerdler/console/blob/13ec0f68fe5a8f37403b67585aad1fc1b2ab89a4/frontend/public/components/cloud-services/index.tsx) has some better type arbitrary examples.

spadgett replied:
`fixed` can really be any... It's just saying always use these properties. I don't think we can define that one.

We can use K8sResourceKind for obj

jhadvig replied:
> Can we eliminate the any types?

no all the types have to be typed, otherwise yarn will screen, eg:
```
Property 'metadata' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<BaseEditSecretProps_, "fixed" | "ki...'
```

jhadvig replied:
Since some parts of codebase are not typed the its almost impossible not to use `any`

stevekuznetsov commented:
/test backend
/test frontend

jhadvig commented:
@spadgett @benjaminapetersen comments addressed.
Regarding the [Higher order composition comment](https://github.com/openshift/console/pull/94#discussion_r196497331) I'm still looking into a way how that could be done, but not sure if it will be a benefit for us.

spadgett commented:
> Regarding the Higher order composition comment I'm still looking into a way how that could be done, but not sure if it will be a benefit for us.

OK to leave for now, but it's something that we might explore before adding all of the dialogs when it's harder to change.

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,151 @@
+import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
+
+import { WEBHOOKSECRETKEY } from '../../const';
```
I would probably just export it in this file.

```js
export const WebHookSecretKey = 'WebHookSecretKey';
```

spadgett commented on frontend/public/components/secrets/subforms.tsx:
```diff
@@ -0,0 +1,49 @@
+import * as React from 'react';
```
These components are going to be so tightly coupled to create secret, I would keep them in the same file. Then you don't need to export them since they're internal to the implementation.

spadgett commented on frontend/public/components/secrets/subforms.tsx:
```diff
@@ -0,0 +1,49 @@
+export class WebHookSecretSubform extends React.Component<WebHookSecretSubformProps, WebHookSecretSubformState> {
+  constructor(props) {
+    super(props);
+    this.state = {WebHookSecretKey: this.props.stringData.WebHookSecretKey || ''};
```
nit: I'd call this `webhookSecretKey`. State properties usually begin with lowercase letters

spadgett commented on frontend/public/components/utils/button-bar.jsx:
```diff
@@ -21,6 +21,7 @@ const InfoMessage = ({message}) => <div className="alert alert-info"><span class
 // NOTE: DO NOT use <a> elements within a ButtonBar.
 // They don't support the disabled attribute, and therefore
 // can't be disabled during a pending promise/request.
+/** @type {React.SFC<{children: any, className?: string, errorMessage?: string, infoMessage?: string, inProgress: boolean}}>} */
```
👍

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -0,0 +1,151 @@
+
+/* eslint-disable no-undef */
+export type BaseEditSecretState_ = {
+  secretType?: string,
```
This would be better as an enum

beanh66 commented:
@jhadvig Do you mind adding a screenshot of the latest? Thanks!

jhadvig commented:
@beanh66 there haven't been much of a changes, just wording in the form:
![1](https://user-images.githubusercontent.com/1668218/41658576-164cbb32-7497-11e8-8a31-95ed18bc36e2.png)
![2](https://user-images.githubusercontent.com/1668218/41658577-16680388-7497-11e8-8a1b-24ff848fcf90.png)
![3](https://user-images.githubusercontent.com/1668218/41658579-1684efa2-7497-11e8-8b84-b5c87ba8999f.png)

beanh66 commented:
@jhadvig Thanks! Question...should the last screen say "Edit Secret" as the title? Wondering why we need _Generic_.

jhadvig commented:
@beanh66 sorry for that put accidentally d'n'd old screen there.. the title should reflect the secret type, in this case it should be - `Edit Webhook Secret`, the same for create page - `Create Webhook Secret`

jhadvig commented:
@spadgett comments addressed and comments added. PTAL

spadgett commented:
/lgtm
/retest

I suspect the ci/prow/* jobs are broken again, but let's check

jhadvig commented:
Thanks @spadgett much appreciated ! :)
